aten/src/ATen/native/TensorConversions.cpp

```diff
@@ -563,95 +563,87 @@ Tensor sparse_compressed_to_dense(
-    const auto to_transposed_csr = at::native::_sparse_csr_tensor_unsafe(
-        self.ccol_indices(),
-        self.row_indices(),
-        self.values(),
-        dst_transposed_sizes,
-        self.values().scalar_type(),
-        kSparseCsr,
-        self.values().device());
-    auto dst_transposed = at::zeros(dst_transposed_sizes, self.options().layout(kStrided));
-    dst_transposed.add_(to_transposed_csr);
-    return dst_transposed.transpose(batch_ndim, batch_ndim + 1);
-  }
-  if (self.layout() == kSparseBsr || self.layout() == kSparseBsc) {
+  if (self.layout() == kSparseCsr || self.layout() == kSparseCsc
+      || self.layout() == kSparseBsr || self.layout() == kSparseBsc) {
     Tensor compressed_indices;
     Tensor plain_indices;
     std::tie(compressed_indices, plain_indices) =
         sparse_csr::getCompressedPlainIndices(self);
 
     auto values = self.values();
     Tensor dense = at::zeros(self.sizes(), self.options().layout(kStrided));
-    if (self.dim() == 2) {
-      // Pad shape so we can treat 2-d like batched, we will squeeze out the
-      // phantom batch dim at the end
+
+    if (batch_ndim == 0) {
+      // Pad shape so we can treat non-batched like batched, we will
+      // squeeze out the phantom batch dim at the end
       compressed_indices.unsqueeze_(0);
       plain_indices.unsqueeze_(0);
-      values = values.unsqueeze_(0);
-      dense = dense.unsqueeze_(0);
+      values.unsqueeze_(0);
+      dense.unsqueeze_(0);
     }
-    if (self.dim() > 3) {
+    if (batch_ndim > 0) {
       // Flatten batch dims
       compressed_indices = compressed_indices.flatten(0, batch_ndim - 1);
       plain_indices = plain_indices.flatten(0, batch_ndim - 1);
       values = values.flatten(0, batch_ndim - 1);
       dense = dense.flatten(0, batch_ndim - 1);
     }
 
-    // At this point everything has 3d shape either the batch dim was inserted,
-    // existed already or was flattened from multiple batch dims
-    std::array<int64_t, 2> blocksize = {values.size(-2), values.size(-1)};
+    // At this point everything the batch dim was inserted, existed
+    // already or was flattened from multiple batch dims
     auto n_batch = values.size(0);
-    // If we already had batch dim(s) and any of them were zero we can take the
-    // early exit.
+    // If we already had batch dim(s) and any of them were zero we can
+    // take the early exit
     if (n_batch == 0) {
       return dense.reshape(self.sizes());
     }
-    // Due to early exit above this reshape should always be valid
-    dense = dense.reshape({n_batch, -1, values.size(-2), values.size(-1)});
+
+    int64_t nrows, ncols;
+    std::array<int64_t, 2> blocksize = {1, 1};
+    if (self.layout() == kSparseCsr || self.layout() == kSparseCsc) {
+      nrows = self.size(batch_ndim);
+      ncols = self.size(batch_ndim + 1);
+
+      dense = dense.flatten(1, 2);
+    }
+    else {
+      blocksize = {values.size(batch_ndim + 2), values.size(batch_ndim + 3)};
```

Comment:
`values` was reshaped above so it contains exactly one batch dimension, but `batch_ndim` tracks the number of batch dims in the original shape.

After the `flatten`/`unsqueeze` of is applied to `values` the generalized shape would be `(n_batch, nnz) + blocksize + dense_dims` where `n_batch`, and `nnz` are scalars, `blocksize` is a length 2 tuple and `dense_dims` is a tuple with as many members as there are dense dimensions in the tensor, possibly zero. 

So at this point the blocksizes can always be computed as `values.size(2)` and `values.size(3)` for any input which enters this region.

Reply:
Fixed - thanks!  Also updated, in the next commit below, according to other comments of yours.